src/sandstorm/grain.capnp

```diff
@@ -23,80 +23,165 @@ using Util = import "util.capnp";
+  # - A powerbox "query" is a list of descriptors used in a request to indicate what kinds of
+  #   objects the requesting app is looking for. (In a powerbox "offer" interaction, the "query"
+  #   is the list of descriptors that the accepting app indicated it accepts in its `ViewInfo`.)
+  # - A powerbox "provision" is a list of descriptors used to describe what kinds of objecst an
+  #   app provides, which can be requested by other apps. (In a powerbox "offer" interaction, the
+  #   "provision" is the list of descriptors that the offering app passed to `offer()`.)
+  #
+  # For a query to match a provision, at least one descriptor in the query must match at least one
+  # descriptor in the provision (with an acceptable `matchQuality`; see below).
+
+  tags @0 :List(Tag);
+  # List of tags. For a query descriptor to match a provision descriptor, every tag in the query
+  # must be matched by at least one tag in the provision. If the query tags list is empty, then
+  # the query is asking for any capability at all; this occasionally makes sense in "meta" apps
+  # that organize or communicate general capabilities.
+
+  struct Tag {
+    id @0 :UInt64;
+    # A unique ID naming the tag. All such IDs should be created using `capnp id`.
     #
-    # Titles are suggestions only.  The user may override the title with their own.
-
-    verbPhrase @1 :Util.LocalizedText;
-    # Verb phrase describing what the holder of this capability can do to the grain, e.g.
-    # "can edit".  This may be displayed in the sharing UI to describe a connection between two
-    # grains.
-
-    description @2 :Util.LocalizedText;
-    # Long-form description of what the capability represents.  Should be roughly a paragraph that
-    # could be displayed e.g. in a tooltip.
-
-    interfaces @3 :List(UInt64);
-    # Type IDs of Cap'n Proto interfaces implemented by this capability.  Only interfaces which
-    # should be used for powerbox searching/matching purposes need to be listed.  Superclasses
-    # must be listed explicitly if they should be considered for matching.  The powerbox does not
-    # know anything about the actual interface schemas; it just matches the IDs.
+    # If `id` is the Cap'n Proto type ID of an interface, it indicates that the described powerbox
+    # capability will implement this interface. The interface's documentation may define what
+    # `value` should be in this case; otherwise, it should be null.
+    #
+    # If `id` is the type ID of a struct type, then `value` is an instance of that struct type.
+    # The struct type's documentation describes how the tag is to be interpreted.
+    #
+    # Apps may also define `typeId` values which do not correspond to any Cap'n Proto type. In
+    # this case, the ID should be declared as a constant with documentation defining what the
+    # `value` should be in this case. For example, you might declare a tag type ID which is for
+    # strings.
```

Comment:
> > you might declare a tag type ID which is for strings.
>
> If we had a `tag.type` field rather than a `tag.id` field (as I suggest above), there would be no need for such ad-hoc declarations.

I disagree. The problem is, the tag ID doesn't just name a type, it identifies the tag _purpose_. If you simply say that the value is Text, what does it mean? When would you ever match on `Text`? In fact there may need to be many different tag IDs which have the type `Text`, each with some specific meaning.